src/extractor/guidance/sliproad_handler.cpp

```diff
@@ -29,273 +32,720 @@ SliproadHandler::SliproadHandler(const IntersectionGenerator &intersection_gener
-        return !node_based_graph.GetEdgeData(road.eid).roundabout && road.entry_allowed &&
-               angularDeviation(road.angle, STRAIGHT_ANGLE) <= 2 * NARROW_TURN_ANGLE &&
-               !hasRoundaboutType(road.instruction) &&
-               angularDeviation(next_road.angle, road.angle) >
-                   std::numeric_limits<double>::epsilon();
-    };
+        // Narrow turn angle for road (bd) and guard against data issues (overlapping roads)
+        auto is_narrow = angularDeviation(road.angle, STRAIGHT_ANGLE) <= 2 * NARROW_TURN_ANGLE;
+        auto same_angle = angularDeviation(next_road.angle, road.angle) //
+                          <= std::numeric_limits<double>::epsilon();
 
-    bool hasNarrow =
-        std::find_if(intersection.begin(), intersection.end(), linkTest) != intersection.end();
-    if (!hasNarrow)
-        return intersection;
+        if (!is_narrow || same_angle)
+            return false;
+
+        // Prevent from starting in or going onto a roundabout
+        auto onto_roundabout = hasRoundaboutType(road.instruction);
 
-    const auto source_edge_data = node_based_graph.GetEdgeData(source_edge_id);
-    // check whether the continue road is valid
-    const auto check_valid = [this, source_edge_data](const ConnectedRoad &road) {
-        const auto road_edge_data = node_based_graph.GetEdgeData(road.eid);
-        // Test to see if the source edge and the one we're looking at are the same road
+        if (onto_roundabout)
+            return false;
 
-        const auto same_name = !util::guidance::requiresNameAnnounced(
-            road_edge_data.name_id, source_edge_data.name_id, name_table, street_name_suffix_table);
+        const auto road_data = node_based_graph.GetEdgeData(road.eid);
 
-        return road_edge_data.road_classification == source_edge_data.road_classification &&
-               road_edge_data.name_id != EMPTY_NAMEID && source_edge_data.name_id != EMPTY_NAMEID &&
-               same_name && road.entry_allowed;
+        auto is_roundabout = road_data.roundabout;
+
+        if (is_roundabout)
+            return false;
+
+        return true;
     };
 
-    if (!check_valid(next_road))
+    if (!std::any_of(begin(intersection), end(intersection), is_potential_link))
         return intersection;
 
-    // Threshold check, if the intersection is too far away, don't bother continuing
-    const auto coordinate_extractor = intersection_generator.GetCoordinateExtractor();
-    const auto next_road_length = util::coordinate_calculation::getLength(
-        coordinate_extractor.GetForwardCoordinatesAlongRoad(
-            node_based_graph.GetTarget(source_edge_id), next_road.eid),
-        &util::coordinate_calculation::haversineDistance);
-    if (next_road_length > MAX_SLIPROAD_THRESHOLD)
+    // If the intersection is too far away, don't bother continuing
+    if (nextIntersectionIsTooFarAway(intersection_node_id, next_road.eid))
```

Comment:
is this check valid? Looks like you are finding the next intersection below. You need to know about the full distance for this check, don't you? Or are you calculating the path to the next intersection twice? (In that case, it could be considered performance bug)